libraries/standard/json/src/json.c

```diff
@@ -0,0 +1,1014 @@
+    }
+
+    *start = i;
+    return ret;
+}
+
+/**
+ * See json.h for docs.
+ *
+ * Verify that the entire buffer contains exactly one scalar
+ * or collection within optional whitespace.
+ */
+JSONStatus_t JSON_Validate( const char * buf,
+                            size_t length )
+{
+    JSONStatus_t ret;
+    size_t i = 0U;
+
+    assert( ( buf != NULL ) && ( length > 0U ) );
+
+    skipSpace( buf, &i, length );
+    #ifndef JSON_VALIDATE_COLLECTIONS_ONLY
+        if( skipAnyScalar( buf, &i, length ) == true )
+        {
+            ret = JSONSuccess;
+        }
+        else
+    #endif
+    {
+        ret = match( buf, &i, length );
+    }
+
+    if( ( ret == JSONSuccess ) && ( i < length ) )
+    {
+        skipSpace( buf, &i, length );
+
+        if( i != length )
+        {
+            ret = JSONIllegalDocument;
+        }
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Output indexes for the next key-value pair of an object.
+ *
+ * Also advances the buffer index beyond the key-value pair.
+ * The value may be a scalar or a collection.
+ *
+ * @param[in] buf  The buffer to parse.
+ * @param[in,out] start  The index at which to begin.
+ * @param[in] max  The size of the buffer.
+ * @param[out] key  A pointer to receive the index of the key.
+ * @param[out] keyLength  A pointer to receive the length of the key.
+ * @param[out] value  A pointer to receive the index of the value.
+ * @param[out] valueLength  A pointer to receive the length of the value.
+ *
+ * @return true if a key-value pair was present;
+ * false otherwise.
+ */
+static bool nextKeyValuePair( const char * buf,
+                              size_t * start,
+                              size_t max,
+                              size_t * key,
+                              size_t * keyLength,
+                              size_t * value,
+                              size_t * valueLength )
+{
+    bool ret = true;
+    size_t i = *start, save = *start;
+
+    if( skipString( buf, &i, max ) == true )
+    {
+        *key = save + 1U;
+        *keyLength = i - save - 2U;
+    }
+    else
+    {
+        ret = false;
+    }
+
+    if( ret == true )
+    {
+        skipSpace( buf, &i, max );
+
+        if( ( i < max ) && ( buf[ i ] == ':' ) )
+        {
+            i++;
+            skipSpace( buf, &i, max );
+            save = i;
+        }
+        else
+        {
+            ret = false;
+        }
+    }
+
+    if( ret == true )
+    {
+        if( ( skipAnyScalar( buf, &i, max ) == true ) ||
+            ( match( buf, &i, max ) == JSONSuccess ) )
+        {
+            *value = save;
+            *valueLength = i - save;
+        }
+        else
+        {
+            ret = false;
+        }
+    }
+
+    *start = i;
+    return ret;
+}
+
+/**
+ * @brief Find a key in a JSON object and output a pointer to its value.
+ *
+ * @param[in] buf  The buffer to search.
+ * @param[in] max  size of the buffer.
```

Comment:
is there any semantic difference between the max vs. length names? it looks to me like just an inconsistency, but maybe I'm missing something

Reply:
No difference.  I made all the names the same.